### ISSUE TITLE
`am` CLI hard-refuses every mailbox subcommand (writes AND reads) while a normal `serve-http` daemon owns the mailbox — no daemon proxy/attach path

### ISSUE THREAD

**Issue by bitjson** [Dicklesworthstone/mcp_agent_mail_rust#126]
(Opus 4.7:)

> ## Summary
> 
> When an `am serve-http` daemon is running (the normal state for any
> multi-agent setup), the `am` CLI does not proxy through it. Every mailbox
> subcommand opens the SQLite store directly, hits the mailbox-ownership
> guard, and hard-refuses:
> 
> ```
> error: database error: SQLite error: mailbox mutation refused for <db>: another Agent Mail process already owns the mailbox: PID <srv> [storage_lock,sqlite_lock,db_file] cmd=... am serve-http ... ; wait for the active owner to finish instead of competing recovery
> ```
> 
> Two problems:
> 
> 1. **Read-only subcommands are refused too.** `am agents list` fails
>    identically to `am agents register`, because the refusal is raised at
>    DB-open / archive-reconcile time, *before* the command is classified as
>    read vs write. A read should never be blocked by a write-ownership guard.
> 2. **There is no daemon proxy/attach path.** The error text says *"Route
>    writes through that process or stop it first"*, but the CLI has no mode
>    that detects and forwards to a running `serve-http` daemon. So whenever
>    the daemon is up, the CLI is unusable and users must hand-roll raw
>    JSON-RPC `curl` calls against the HTTP endpoint.
> 
> Affects `am` 0.2.51 (latest release). Verified present at `main` HEAD
> `bc52c5be15597bdd53231e4151c8e0ee78aeb817` — the implicated functions are
> unchanged from `v0.2.51`.
> 
> ## Reproduction
> 
> Self-contained; isolated `STORAGE_ROOT` + private port, no global config
> touched. (Full script available as a reproduction artifact.)
> 
> ```bash
> ROOT=$(mktemp -d); export STORAGE_ROOT="$ROOT"
> export DATABASE_URL="sqlite+aiosqlite:///$ROOT/storage.sqlite3"
> PORT=8809; U="http://127.0.0.1:$PORT/mcp/"
> 
> # 1. start daemon, wait until /health reports "status":"ready"
> am serve-http --no-tui --no-auth --host 127.0.0.1 --port $PORT --path /mcp/ &
> until curl -s "http://127.0.0.1:$PORT/health" | grep -q '"status":"ready"'; do sleep 0.5; done
> 
> # 2. register ONE agent THROUGH the daemon so the git archive commits
> #    <ROOT>/projects/<slug>/ (this arms the ownership guard; a live
> #    deployment always satisfies this)
> curl -s -X POST "$U" -H 'Content-Type: application/json' \
>   -H 'Accept: application/json, text/event-stream' \
>   -d "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"tools/call\",\"params\":{\"name\":\"register_agent\",\"arguments\":{\"project_key\":\"$ROOT\",\"program\":\"claude-code\",\"model\":\"t\",\"task_description\":\"seed\"}}}"
> until [ -n "$(ls $ROOT/projects 2>/dev/null)" ]; do sleep 0.5; done
> 
> # 3-5. CLI subcommands while the daemon owns the mailbox:
> am agents register --project "$ROOT" --program claude-code --model test --task w   # REFUSED
> am agents create   --project "$ROOT" --program claude-code --model test --task w2  # REFUSED
> am agents list     --project "$ROOT"                                               # REFUSED (read!)
> 
> # 6. the same mutation via the daemon's JSON-RPC succeeds — proving the asymmetry
> curl -s -X POST "$U" -H 'Content-Type: application/json' \
>   -H 'Accept: application/json, text/event-stream' \
>   -d "{\"jsonrpc\":\"2.0\",\"id\":9,\"method\":\"tools/call\",\"params\":{\"name\":\"register_agent\",\"arguments\":{\"project_key\":\"$ROOT\",\"program\":\"claude-code\",\"model\":\"test\",\"task_description\":\"via-http\"}}}"
> ```
> 
> Expected: steps 3–5 succeed (proxied through the daemon, or at minimum the
> read in step 5 succeeds). Actual: steps 3, 4, **and 5** all fail with the
> refusal above; step 6 succeeds.
> 
> ## Root cause
> 
> - `crates/mcp-agent-mail-db/src/pool.rs:105-110` —
>   `MailboxOwnershipState::blocks_mutation()` returns `true` for every
>   disposition except `Unowned`; a healthy warmed `serve-http` daemon is
>   `ActiveOtherOwner` (`pool.rs:5516-5554`).
> - `crates/mcp-agent-mail-db/src/pool.rs:5632-5652` —
>   `refuse_mutating_mailbox_when_owned` emits the observed string. Its
>   callers (`pool.rs:3701`, `6200`, `6733`) sit on the DB-open /
>   archive-reconcile path (`reconcile_archive_state_before_init`,
>   `pool.rs:3690`), so the refusal precedes the read/write split — that is
>   why read-only subcommands are blocked.
> - `crates/mcp-agent-mail-db/src/pool.rs:1573-1612` — `evaluate_write_route`
>   already prescribes *"Route writes through that process"*, but the CLI
>   has no router: `handle_mail` /
>   `handle_agents`→`handle_agents_async`
>   (`crates/mcp-agent-mail-cli/src/lib.rs:24886`, `25899`) open the SQLite
>   store directly with no `serve-http`-detection / HTTP-forward branch.
> 
> ## Suggested fix shape
> 
> 1. **Don't block reads on a write-ownership guard.** Move
>    `refuse_mutating_mailbox_when_owned` off the unconditional DB-open path,
>    or pass a read/write intent into `reconcile_archive_state_before_init`
>    so read-only subcommands open a read-only / archive-snapshot view even
>    when another process owns the mailbox.
> 2. **Implement the CLI→daemon proxy the error message already promises.**
>    When an `ActiveOtherOwner` daemon is detected, have mailbox subcommands
>    resolve the daemon's HTTP endpoint + bearer token (the MCP client config
>    the project already writes) and forward the operation as JSON-RPC
>    instead of contending for the SQLite lock — symmetrical with what
>    external clients already do successfully (step 6). This is the same
>    `is_authority`/route-through concept already present in
>    `evaluate_write_route`, extended to an actual transport.
> 
> ## Workaround (today)
> 
> Bypass the CLI entirely while the daemon is up: POST JSON-RPC to the
> daemon's `/mcp/` endpoint with its bearer token (e.g.
> `tools/call` `register_agent`, `send_message` with `body_md`). Brittle and
> unvalidated, but the only path that works concurrently with the daemon.
> 
> ## Severity / scope
> 
> High for multi-agent / swarm use: the daemon is the normal running state,
> so the CLI is effectively unusable for both writes and reads in that mode,
> and every agent must hand-roll raw HTTP. Single-user, daemon-off use is
> unaffected. A related but distinct bug (intermittent `resources/read …
> request timed out` on `list_agents` under concurrent load, caused by an
> uncached per-request archive-snapshot reconstruction in
> `crates/mcp-agent-mail-tools/src/resources.rs:325-417`) was also observed
> and is being reported separately. Reproduction artifacts available.

**Comment by Dicklesworthstone** [comment]
Thanks — this is an exceptionally well-localized report. I walked the paths you cited and your root-cause analysis holds up exactly:

- `MailboxOwnershipState::blocks_mutation()` (`pool.rs:105-110`) treats every disposition except `Unowned` as mutation-blocking, and a warm `serve-http` daemon is `ActiveOtherOwner`.
- `refuse_mutating_mailbox_when_owned` fires from the archive-reconcile / DB-open path (`reconcile_archive_state_before_init`) *before* the command is classified read vs. write, which is precisely why `am agents list` is refused identically to a write.
- `evaluate_write_route` already names the intended remedy ("route writes through that process") but no CLI router exists; `handle_mail` / `handle_agents_async` open SQLite directly.

I'm treating this as two separable pieces of work:

1. **Reads must never be blocked by a write-ownership guard.** This is a clear correctness bug and the smaller of the two — the fix is to defer the ownership refusal until after the read/write intent is known, and let read-only subcommands open a read-only snapshot view even when the daemon owns the mailbox. This is the part I'd land first.
2. **Daemon proxy/attach path for writes.** Forwarding mutating subcommands to the running `serve-http` JSON-RPC endpoint (reusing the config the project already writes, with bearer-token/port discovery) is the right end state but a larger design effort — it has to handle endpoint discovery, auth, and graceful fallback when no daemon is up.

Leaving this open as the canonical tracker for both. No ETA, but the read-path asymmetry is the kind of thing that shouldn't linger, so it's prioritized. Appreciate the self-contained repro — it makes validating the fix straightforward.

**Comment by Dicklesworthstone** [comment]
Part **(a)** — the read-classification fix — has shipped on `main` in 8bd07613.

The fix follows the maintainer's suggested shape (1) directly:

- `mcp-agent-mail-db::pool` now exposes a per-thread `ReadOnlyIntentGuard` RAII guard. While at least one guard is alive on the current thread, two write-only behaviours in the DB-init path are suppressed:
  - `refuse_mutating_mailbox_when_owned()` returns `Ok(())` early — a read never mutates, so a write-owner is irrelevant.
  - `reconcile_archive_state_before_init()` returns `Ok(false)` early — reconstruction *is* a mutation that would contend with the owning daemon's WAL.
- The CLI dispatcher (`mcp-agent-mail-cli::execute`) classifies each subcommand and enters the guard before the handler opens the pool when the classification is read-only. Subcommands default to write-classified so the change is conservative.

Read-classified subset that covers the reporter's repro (`am agents list`):

- Top-level: `inbox`, `thread`, `health`, `check-inbox`, `status`, `capabilities`, `reservations`, `list-projects`, `list-acks`.
- `agents`: `list`, `show`, `detect`.
- `mail`: `status`, `inbox`, `read`, `search`, `summarize-thread`.
- `acks`: `pending`, `overdue`.
- `file-reservations`: `list`, `active`, `soon`, `conflicts`.
- `contacts`: `list-contacts`.

Unit-test coverage in `pool::tests` (all 4 green):

- `read_only_intent_guard_nests_and_drops` — guard depth semantics.
- `read_only_intent_is_thread_local` — flag does not leak between threads.
- `reconcile_archive_state_before_init_short_circuits_under_read_intent` — the primary regression: a primed archive that would normally drive a reconstruct returns `Ok(false)` under the guard and leaves the primary file untouched.
- `refuse_mutating_mailbox_when_owned_passes_under_read_intent` — pins the early-return path so a future tightening would fail this test.

`cargo check -p mcp-agent-mail-db -p mcp-agent-mail-cli` green; `cargo fmt --check` green.

One caveat that lands inside part (a)'s scope: if the primary file is missing AND read intent is set, the open will still fail (no DB to read from). That case is what part **(b)** — the daemon-proxy/attach path that forwards reads as JSON-RPC through the live `serve-http` daemon — would solve. **Part (b) is intentionally left open** here; it's the larger multi-day design that needs HTTP/MCP client wiring, endpoint discovery from the project's MCP config, bearer-token plumbing, and a transparent fallback when no daemon is running. Will track it as a separate issue once the design lands.

**Comment by Dicklesworthstone** [comment]
Fixed and verified on current `main`.

What changed:

- CLI mailbox commands now proxy through the running daemon first for the high-value paths that caused this report (`agents register/create/list/show`, `whois`, and `macro start-session`) instead of competing with the live SQLite owner.
- HTTP JSON-RPC error handling now preserves structured 401/403/429 envelopes instead of collapsing them into opaque transport failures.
- Static bearer auth now carries write authority, and `list_agents` is allowed as a read-only RBAC path.
- Local daemon endpoint/token discovery now includes managed service config fallbacks.
- The HTTP E2E suite now has a live-daemon CLI proxy run that exercises the exact failure mode.

Proof run today (2026-05-23):

```text
cargo check --workspace --all-targets                         # pass
cargo clippy --workspace --all-targets -- -D warnings          # pass
cargo test --workspace                                        # pass
cargo fmt --check                                             # pass

timeout 1800 env CARGO_TARGET_DIR=/data/projects/mcp_agent_mail_rust/target \
  AM_E2E_HTTP_REQUIRE_PASS=1 AM_E2E_HTTP_INCLUDE_FOCUSED_SUBSUITES=1 \
  cargo test -p mcp-agent-mail-cli --test http_transport_harness -- --nocapture
# pass; includes live daemon CLI proxy/send_message survival coverage
```

Live daemon proof after deploying the rebuilt binaries:

- local host: `am agents show --project /data/projects/frankensqlite WildGrove --json` succeeds while `agent-mail.service` owns the mailbox; `am macros start-session ...` created `MaroonPond`; `am mail send ...` returned message `id=4783`; readiness stayed green.
- `ts2`: `am agents list --project /data/projects/hedge_fund_data_tool --json | jq length` returned `346`; `am macros start-session ...` created `GoldTrout`; `am mail send ...` returned message `id=17491`; readiness stayed green.

This resolves both sides of the report: reads no longer hard-refuse behind a live daemon, and the practical write paths agents use now attach/proxy through the daemon instead of directly contending for the mailbox.
